blogposts/2020/code-search-turned-code-checker.md

```diff
@@ -0,0 +1,296 @@
+information and variable scope to implement checks. At the same time,
+`staticcheck` isn't a search tool, it's an entire toolset that includes a suite
+of pre-written, high-precision checks that's very effective in certain
+workflows, like CI. The question is not necessarily whether a search tool can
+achieve parity with a tool like `staticcheck`. But given the overlap with
+now-expressible search queries, it does prompt: how far can we push structural
+code search to find similarly _actionable_ code checks? I.e., checks that match
+real cases of code that we can improve.
+
+So, taking inspiration from `staticcheck`, I wanted to see how many of its checks
+translate to search queries that I could have high confidence in (i.e., all
+patterns find legitimate issues; zero or very-near-zero false positives). I
+chose to look at `staticcheck` checks for its clear documentation, which made it
+easy to start developing checks. So I attempted to write search patterns
+for each [simple static check](https://staticcheck.io/docs/checks).<sup>3</sup>
+
+### Test files as reference patterns
+
+I ran my search queries against `staticcheck`'s own test files to check that they
+don't match unintended patterns (false positives) and don't miss real patterns
+(false negatives). Each check may have more than one syntactic _variant_, so I
+tried to implement patterns for as many variants as I could find in tests. It's
+a neat exercise to develop patterns against the reference tests and discover
+which variants to cover, all in a self-contained search webapp. Here's an example
+where the query matches all the true hits in the test file, annotated with
+`// want strings.Contains ...`:
+
+
+<img src="https://storage.googleapis.com/sourcegraph-assets/about.sourcegraph.com/blog/2020/multiline-query-editor.png">
+
+[🔘 Example query to match known patterns in test files↗](https://sourcegraph.com/search/console?q=repo:%5Egithub%5C.com/dominikh/go-tools$%20%0Alang:go%0Afile:test%0A%0Astrings.IndexRune(...,%20...)%20%3E%20-1%0A%0Aor%0A%0Astrings.IndexRune(...,%20...)%20%3E=%200%0A%0Aor%0A%0Astrings.IndexRune(...,%20...)%20!=%20-1%0A%0Aor%0A%0Astrings.IndexRune(...,%20...)%20==%20-1%0A%0Aor%0A%0Astrings.IndexRune(...,%20...)%20%3C%200%0A%0Aor%0A%0Astrings.IndexAny(...,%20...)%20%3E%20-1%0A%0Aor%0A%0Astrings.IndexAny(...,%20...)%20%3E=%200%0A%0Aor%0A%0Astrings.IndexAny(...,%20...)%20!=%20-1%0A%0Aor%0A%0Astrings.IndexAny(...,%20...)%20==%20-1%0A%0Aor%0A%0Astrings.IndexAny(...,%20...)%20%3C%200%0A%0Aor%0A%0Astrings.Index(...,%20...)%20%3E%20-1%0A%0Aor%0A%0Astrings.Index(...,%20...)%20%3E=%200%0A%0Aor%0A%0Astrings.Index(...,%20...)%20!=%20-1%0A%0Aor%0A%0Astrings.Index(...,%20...)%20==%20-1%0A%0Aor%0A%0Astrings.Index(...,%20...)%20%3C%200%0A)
+
+Using the test files isn't a guarantee that I've implemented all the checks or
+that it's entirely precise, but it adds a lot more confidence than inferring
+patterns only from documentation.
+
+### Results
```

Comment:
I think it would be interesting to show a small table comparing the definitions of these checks in `staticcheck` (presumably long/complex) vs. comby.